_posts/2015-11-16-CodeRefacorting.md

```diff
@@ -0,0 +1,24 @@
+
+Sometimes refactoring code is not enough to deal with outdated code. Typically refactoring code does not change the functionality of the program, but as any developer can tell you, a project's requirements will always change over time, therefore the code will need to be reworked! In this case "reworking" is just a broader term for refactoring. Refactoring is focused on improving the quality of the code itself, while keeping the functionality the same. Reworking can be anything from completely rewriting code to taking existing code and tweaking it to get a new desired functionality.
+
+Refactoring and reworking code are both essential to keeping a good code base and having a functioning program, but are used at different times. For example, in an open source project such as Waterbear, where contributors are always different, it is very easy for duplicate code to be produced. Sometimes it may be because the code base is large and a new contributor doesn't know where to find a function they need, so they write one instead. Other times it can simply be because they were in a rush and wanted slightly different functionality from an existing implementation so they wrote a new, very similar function. In either case, the duplicate code can be refactored so that there is only one version of the function that encompasses all the previous functions' abilities. But be careful, sometimes this can result in a bloated function! A bloated function is a function that is very long and does many things. Once again this can be solved by refactoring! The bloated function can be separated into multiple functions, each with a specific task, that can then be called to perform the desired functionality. With this sort of refactoring, there is no change in functionality but the code has been changed to be more organized and easier to use and maintain in the future!
+
+For an example of reworking code I will give an example of a recent change to the Waterbear language. As a new contributor to the project I was making an example with Waterbear to try and get familiar with it. As I was doing so, I can across some blocks that I could not figure out how to use, the `Path` blocks. One of the many nice things about open source projects is that there are constantly new, fresh eyes looking at the program. As new contributors are added to a project they are able to see issues in the code or program, that the other developers may have never noticed. Sometimes when developers are too close to a project they fail to see when the software doesn't make sense for new users. As was the case with the `Path` blocks. To the more seasoned developers, the `Path` blocks were easy enough to use, to me they were so complicated that I had to ask how they functioned before I was able to draw a path. When this sort of thing happens it is often a good indication that the code does not make sense and should be reworked!
```

Comment:
The Path blocks were never easy to use, we just hadn't gotten around to figuring out how to fix them yet. Many of the WB blocks start off modeled very close to how JavaScript or the Canvas API wok. Then, as we use them we figure out ways to make things better.